#Introduction#

With the development of next generation sequencing technologies, analyses and functional prediction of sequences show increasingly important. Alternative splicing, a post-transcriptional modification, occurs in pre-mRNA. It can regulate a series of biological processes and play important roles in disease and development. Therefore, the prediction of alternative splicing has become a popular research direction. 
Due to the diversity of biological evolution and has a lot of genetic information, genomic data display a high degree of complexity. Deep learning methods are suitable of this high-dimensional multi-noise data. However, deep learning is rarely used in biological problems. 
We innovatively used convolutional neural networks and recurrent neural network algorithms. To explore the sequence signature of alternative splicing, we originally designed and trained four deep models: splice site predication model, exon reconstruction model, adjacent exon connection model and transcripts reconstruction model. 
Our results show that all four models have good performances at predicting unknown transcripts. We explored the potential of deep learning in alternative splicing in the bioinformatics, indicating new ideas and methods for other researchers.
